### DISCUSSION TITLE
Would it be possible to capture the comments in the AST?

### DISCUSSION THREAD

**Discussion by alexaandru** [deepnoodle-ai/risor#195] [Ideas]
At a quick glance I could not find them in there, which makes sense, the VM doesn't need them.

However, the AST would also be very valuable for writing a tool for auto formatting, and currently, such a tool would lose the comments information, which obviously would be unacceptable.

**Comment by myzie** [comment]
Hi @alexaandru, thanks for pointing this out. It does sound like something we should do. I'll create a linked issue for this.

I'm definitely open to volunteers to help with the implementation :-)

**Reply by alexaandru** [reply]
Awesome & will keep that in mind :-) I have very little spare time these days so can't really promise anything, but if I were to start, I'd probably start here: https://git.sr.ht/~technomancy/fnlfmt/tree/main/item/fnlfmt.fnl and see how that is working :-)

That's the auto-format tool for Fennel lang, it's really tiny & clean and also based on an AST. Can quickly give an idea of the type of information needed (as comments come in quite a few shapes and sizes: top level/"floating" (not attached to any statement), "attached" to the next statement (like the comment you place right above a function or any statement really), trailing, etc.). And somehow all the information needs to be encoded into the AST.

There's also https://github.com/golang/go/tree/master/src/cmd/gofmt of course, but that seems to be a lot more code than in fnlfmt's case :-)

**Comment by myzie** [comment]
I took a stab or two at this, but it turned into a bigger effort than I had hoped. I'm going to close this for now since it's beyond what I can commit to currently.
